Improving Variable Elimination (VE)

jhonatanoliveira opened the pull request:
Here, we finish the implementation of VE with few missing steps, namely **computing good elimination orderings** (with 4 heuristics - _min neighbors_, _min fill_, _min weight_, and _weighted min fill_) and **safely removing irrelevant variables from the model** (_barren_ and _independent by evidence_ nodes).
In order to make these improvements, few new methods were necessary and modifying other members too. In our preliminary experimental results, queries that were taking up to 30 minutes using VE now take less than 2 minutes. Please, help us test this new code for a robust implementation and also looking for better ways of coding the algorithms. Thanks.

jhonatanoliveira committed:
A temporary solution for BIF reading files

A very rough class for extracting a Bayesian Model from a BIF file.

jhonatanoliveira committed:
Optimizing the read BIF file

jhonatanoliveira committed:
Implementing "get descendants" and making "get ancestors" public

jhonatanoliveira committed:
Implemented inaugural variables

An algorithm to return all inaugurals in a DAG, given variables
(usually from a test of independence, say I(X,Y,Z)).

jhonatanoliveira committed:
Fill in edges

Used during the computation of elimination orderings: they are the
edges needed to be added to the graph, when a variable is removed.

jhonatanoliveira committed:
Get cardinality of a variable from a Bayesian Model

Notice that the only way to get the cardinality of a variable is using
the CPD, but it can be useful to get the cardinality from a Bayesian
Model itself (like when computing elimination orderings). This can be
seen as a shortcut.

jhonatanoliveira committed:
A Elimination Ordering class

A class to deal with finding good elimination ordering. Contains 4
heuristics for scoring variables in a moralized graph.

jhonatanoliveira committed:
Variable Elimination computes a good elimination ordering

If no elimination ordering is given, VE uses the heuristic Weighted Min
Fill to compute a good elimination ordering. But to do that, VE needs
access to the model, thus the model is saved in parent class, Inference.

jhonatanoliveira committed:
Wrong variables were sent to elimination ordering

According to the documentation, “variables” should be the ones you want
the posterior, and not the ones to be eliminated. But that’s not what
was happening.

jhonatanoliveira committed:
Move ancestors and descendants to Directed Graph

Also, a helper method for computing leaf nodes.

jhonatanoliveira committed:
Remove independent by evidence variables

Fully implemented Variable Elimination, with barren and independent
variables being removed. (Inaugural variables is not fully implemented
yet). Barren and independent are computed by a static method, since
they prune the model, creating sub-models.

jhonatanoliveira committed:
Removing "new root variables"

Also: a roots method.

jhonatanoliveira committed:
typo

jhonatanoliveira committed:
Improving code to increase the health of the repository

jhonatanoliveira committed:
Last changes to improve health of the source code.

palashahuja commented:
@jhonatanoliveira  I am not really sure if this is of any use but please have a look. This method is quite different than the usual one. Here is the [link](http://www.aaai.org/ocs/index.php/AAAI/AAAI14/paper/download/8596/8512) . Please correct me if I am wrong.

jhonatanoliveira commented:
Thanks for the comment, @palashahuja. I didn't know this paper and it's a very interesting approach, but for relational models. I assumed that our class _BayesianModel_ may mostly be suitable for representing a Bayesian network (please, correct me if I'm wrong). Given that, as highlighted in this paper you posted, these heuristics I've implemented "are known to produce good orderings". The way I've implemented _EliminationOrdering_ is as proposed on page 314 in "Probabilistic Graphical Models - Principles and Techniques". But anyway, I'll be happy to modify something here in order to improve VE :)

abinashpanda commented on pgmpy/base/UndirectedGraph.py:
```diff
@@ -174,3 +174,29 @@ def is_triangulated(self):
+        for variable in variables:
+            neighbors = self.neighbors(variable)
+            edges = [e for e in itertools.combinations(neighbors, 2) if not self.has_edge(*e)]
+        return edges
```
@jhonatanoliveira Correct me if I am wrong. The method `fill_in_edges` takes input a `list` of nodes to be removed. But by looking into the code it seems that the method would return only the edges to be removed only for the last variable as `edges` variable is updated in each iteration without storing the previous results.

Furthermore could you please write the tests for each of the method that you wrote with all possible test cases. That would make these methods more robust.

abinashpanda commented on pgmpy/inference/EliminationOrdering.py:
```diff
@@ -0,0 +1,101 @@
+        of a v-structures are connected).
+    """
+
+    def __init__(self, bayesian_model):
```
This class becomes very specific to `BayesianModel`. Although its rare to see variable elimination applied to `MarkovModel` but we want our classes and methods to be as generic as possible. So instead of `bayesian_model` use `model` and put explicit checks that if it is a `BayesianModel` then moralize it.

jhonatanoliveira commented:
Hi, @abinashpanda . Sorry the delay. Indeed, that's a mistake in `fill_in_edges`, thanks for spotting that. I'll fix it right away, and I'll make your other proposes (test cases and changing from `bayesian_model` to `model`) as soon as possible.

jhonatanoliveira committed:
Fix fill_in_edges for multiple variables being eliminated

A mistake was assigning only the last result in the for loop, now the
list with the edges is extended.

jhonatanoliveira committed:
Changing variable name to model and checking for a Bayesian model like suggested by @abinashpanda.

jhonatanoliveira committed:
Tests for leaves and roots

jhonatanoliveira committed:
Back to _get_ancestors (no reason to make it public)

jhonatanoliveira committed:
Make a copy of the original model when setting up Inference

The original model can be useful in future, for instance, when
computing Elimination Orderings.

jhonatanoliveira committed:
Test method for fill_in_edges

The example isn’t good, since the BIF reader doesn’t exist within the
library yet.

jhonatanoliveira committed:
Check if the model is BayesianModel and remove static method

No reason for static method. And the checking for BayesianModel wasn’t
properly done before.

jhonatanoliveira committed:
A new test class for EliminationOrdering tests

jhonatanoliveira committed:
Tests for barren and independent by evidence variables. Methods aren't static anymore.

Also, “new root” method was removed, since it’s a simple logic and can
be done locally.

jhonatanoliveira committed:
Get cardinality test and consistency check

A helper function to get the cardinality of a variable from the
BayesianModel. Check if the variable is within the scope of the graph
and of at least one CPD.

jhonatanoliveira committed:
The BIF reader isn't well done yet.

Removing this BIF reader, which was hand crafted only as a short time
solution for reading BIF files. But this version isn’t stable nor
optimized. Will post it in future.

jhonatanoliveira committed:
General improvements in VE and some fixes (more in description)

Fix the not necessary use of new variables. The removal of irrelevant
variables is only defines for BayesianModel when marginalizing. Safe
copy of the model before performing inference. General fixing to agree
with PEP8.

jhonatanoliveira commented:
@abinashpanda I've finally finished the improvements you've proposed. Please, let me know if there is any further improvement on this PR. We'll be happy to help. Regards.

ankurankan commented:
@jhonatanoliveira I will try to review and merge this PR by this weekend.

jhonatanoliveira committed:
EliminationOrdering is only defined for BayesianModels

Add a test to make sure that EliminationOrdering is only called with
BayesianModels.

jhonatanoliveira committed:
Backup Multiplication and Division to the current PgmPy implementation

As shown in comments, the proposed implementation was slower in a
different benchmark setting. Then, we’re backing up to the current
version, until an improvement is done.

jhonatanoliveira committed:
Merge branch 'pgmpy/dev' into dev

# Conflicts:
#	pgmpy/inference/ExactInference.py
#	pgmpy/models/BayesianModel.py

jhonatanoliveira committed:
Fix problem when more than one variable is queried: the final potential was being modified, so "inplace=False" was incorporated.

jhonatanoliveira committed:
Suggestions by Landscape.

jhonatanoliveira committed:
List is modified during its iteration.

jhonatanoliveira commented:
@ankurankan is dealing with this in another PR.
